### ISSUE TITLE
OpPtrEqual without OpCapability VariablePointers

### ISSUE THREAD

**Issue by neurotok** [EmbarkStudios/rust-gpu#741]
Hello my name is Martin and I was assigned to develop a new set of CTS tests for Khronos to cover the use of OpPtrEquale/OpPtrNotEqual without VariablePointers extension. SPIR-V spec does not make it clear that this op code can only be used in conjunction with SPV_KHR_variable_pointers.

Because I'm a huge Rust fan and I'm very excited about what you are trying to achieve here! I decided to run some simple compute shader with pointer comparison and I got this:

**error: OpPtrEqual without OpCapability VariablePointers**

Can I ask what was the reason behind putting such a blockade on a usage of this op code without variable pointers? I am curious what is the relationship between this extension and this op code from your perspective. It would be very valuable insight for me to hear your opinion on this topic.

**Comment by khyperia** [comment]
> SPIR-V spec does not make it clear that this op code can only be used in conjunction with SPV_KHR_variable_pointers.

Unfortunately, it is indeed in the spec. Looking at section 2.16.1, "Universal Validation Rules", under the section "If neither the VariablePointers nor VariablePointersStorageBuffer capabilities are declared, the following rules apply to logical pointer types:", there is the statement:

**The instructions OpPtrEqual and OpPtrNotEqual must not be used.** 

Further along, in the section "If the VariablePointers or VariablePointersStorageBuffer capability is declared, the following are allowed for logical pointer types:", there are some additional constraints and allowances for OpPtrEqual/OpPtrNotEqual, generally allowing those instructions in some restricted cases.

So, in conclusion, OpPtrEqual can only be used if VariablePointers is declared.

**Comment by neurotok** [comment]
Thank you so much Ashley! I will try to raise this topic and mayby I’ll able to clarify the spec. Good luck with the project!, I will be peeking here from time to time :) I'm not an expert on SPIRV (I mainly deals with vulkan CTS test), but if you need any help, I volunteer! :)
